mdloader_common.c

```diff
@@ -377,6 +377,110 @@ int test_mcu(char silent)
+{
+    //Set WMODE to Manual
+    NVMCTRL_CTRLA_Type ctrla;
+    ctrla.reg = read_half_word(NVMCTRL_CTRLA);
+    printf("NVMCTRL.CTRLA: 0x%04x\n\tAUTOWS: 0x%01x\n\tSUSPEN: 0x%01x\n\tWMODE: 0x%02x\n\tPRM: 0x%02x\n\tRWS: 0x%04x\n\tAHBNS0: 0x%01x\n\tAHBNS1: 0x%01x\n\tCACHEDIS0: 0x%01x\n\tCACHEDIS1: 0x%01x\n", ctrla.reg, ctrla.bit.AUTOWS, ctrla.bit.SUSPEN, ctrla.bit.WMODE, ctrla.bit.PRM, ctrla.bit.RWS, ctrla.bit.AHBNS0, ctrla.bit.AHBNS1, ctrla.bit.CACHEDIS0, ctrla.bit.CACHEDIS1);
+    ctrla.bit.WMODE = NVMCTRL_CTRLA_WMODE_MAN;
+
+    printf("Configuring SmartEEPROM... ");
+
+    if (!write_half_word(NVMCTRL_CTRLA, ctrla.reg))
+    {
+        printf("Error setting NVMCTRL.CTRLA.WMODE to Manual\n");
+        return 0;
+    }
+    slp(100);
+
+    // Set user row address
+    if (!write_word(NVMCTRL_ADDR, NVMCTRL_USER))
+    {
+        printf("Error setting NVMCTRL_ADDR to NVMCTRL_USER (1)\n");
+        return 0;
+    }
+
+    // Erase page
+    NVMCTRL_CTRLB_Type ctrlb;
+    ctrlb.reg = 0;
+    ctrlb.bit.CMD = NVMCTRL_CTRLB_CMD_EP;
+    ctrlb.bit.CMDEX = NVMCTRL_CTRLB_CMDEX_KEY;
+    if (!write_half_word(NVMCTRL_CTRLB, ctrlb.reg))
+    {
+        printf("Error setting NVMCTRL_CTRLB to 0x%04x (Erase page)\n", ctrlb.reg);
+        return 0;
+    }
+    slp(100);
+
+    // Page buffer clear
+    ctrlb.reg = 0;
+    ctrlb.bit.CMD = NVMCTRL_CTRLB_CMD_PBC;
+    ctrlb.bit.CMDEX = NVMCTRL_CTRLB_CMDEX_KEY;
+    if (!write_half_word(NVMCTRL_CTRLB, ctrlb.reg))
+    {
+        printf("Error setting NVMCTRL_CTRLB to 0x%04x (Page buffer clear)\n", ctrlb.reg);
+        return 0;
+    }
+    slp(100);
+
+    // Write in the write buffer
+    for (int i = 0; i < 4; i++)
+    {
+        if (!write_word(NVMCTRL_USER + i * 4, data[i]))
+        {
+            printf("Warning: Unable to write NVMCTRL_USER page %i\n", i);
+        }
+        slp(100);
+    }
+
+    if (!write_word(NVMCTRL_ADDR, NVMCTRL_USER))
+    {
+        printf("Error setting NVMCTRL_ADDR to NVMCTRL_USER (2)\n");
+        return 0;
+    }
+    slp(100);
+
+    // Write quad word (128bits)
+    ctrlb.reg = 0;
+    ctrlb.bit.CMD = NVMCTRL_CTRLB_CMD_WQW;
+    ctrlb.bit.CMDEX = NVMCTRL_CTRLB_CMDEX_KEY;
+    if (!write_half_word(NVMCTRL_CTRLB, ctrlb.reg))
+    {
+        printf("Error setting NVMCTRL_CTRLB to 0x%04x (Write Quad Word)\n", ctrlb.reg);
+        return 0;
+    }
+
+    printf("Success!\n");
+    return 1;
+}
+
+void read_user_row(void)
+{
+    uint32_t user_row[4];
+    printf("user row: ");
+    for (int i = 0; i < 4; i++)
+    {
+        user_row[i] = read_word(NVMCTRL_USER + i * 4);
+        printf("0x%08x ", user_row[i]);
+    }
+    printf("\n");
+
+    NVMCTRL_USER_ROW_MAPPING1_Type* puser_row1 = (NVMCTRL_USER_ROW_MAPPING1_Type*)(&user_row[1]);
+    if (puser_row1->bit.SBLK == 0 && puser_row1->bit.PSZ == 0)
+    {
+        printf("SmartEEPROM not configured, proceed\n");
+        puser_row1->bit.SBLK = 0x2; // 2 blocks
+        puser_row1->bit.PSZ = 0x1;  // 8 bytes
```

Comment:
@sigprof This is what I asked in my previous comment.
That would require that https://github.com/qmk/qmk_firmware/pull/6068 is updated, right?
I want to keep backwards compatibility.

How do you suggest that we achieve this?

Reply:
Okay, so I re-read the code and the manual regarding this. I'm setting it now to:

```
    puser_row1->bit.SBLK = 0x1; // 1 block
    puser_row1->bit.PSZ = 0x1;  // 8 bytes
```

According to table 25-6, this will set the virtual size to 1024 bytes, in the optimal size.
Looking at https://github.com/qmk/qmk_firmware/pull/6068, it seems it checks to see if either `SBLK` or `PSZ` is 0 - and if any of them are, disable the functionality.
Looking further into it, it seems the user configuration just requires 36 bytes for storage, so this could probably be set to:

```
    puser_row1->bit.SBLK = 0x1; // 1 block
    puser_row1->bit.PSZ = 0x0;  // 4 bytes
```

Yielding a 512 bytes storage, in optimum setting - but this would require an update to the QMK PR.

For the meantime, I'm changing to code to the first option, but if you think we should push this further to the 2nd option, I can do the change - just waiting on your feedback.

Reply:
One of important users of EEPROM is VIA, and 1024 bytes is a commonly used EEPROM size (e.g., the ATmega32U4 chip (probably the most popular MCU for custom keyboards) has 1024 bytes of EEPROM). 1024 bytes is usually enough to store 4 keymap layers for a TKL keyboard, and going below that would be problematic. As for going to 2048 bytes of EEPROM instead, I'm not really sure — this would allow fitting more VIA keymap layers while occupying the same amount of flash as SBLK=2, PSZ=1, but maybe page rewrites for the settings part would happen more often because of larger page size. However, actually using the extra EEPROM space would require some changes in the QMK code (the current PR does not even try to support VIA, providing support just for a minimal amount of EEPROM required for the QMK core).

And actually the table gives 512 bytes as an optimum size for SBLK=1, PSZ=0 only because there is no way to go any lower than SBLK=1 — some flash space is still wasted in this configuration, so there is probably no point in going that low.

Reply:
@sigprof Fair enough, looks good for this PR, then.

Still, I didn't know about VIA existence ([this one, right?](https://caniusevia.com/)), so might be a good idea to plan ahead for it.
Where can I get more info about VIA and/or how to future proof this PR to support it?

I can happily create a new PR in qmk repo to support it, if you give me some advice on how to proceed :)

Reply:
From the EEPROM side of things VIA just needs a large enough EEPROM (1024 bytes is enough, although 2048 might be better for a layout larger than TKL) with fast enough read access (every keymap lookup generates several EEPROM reads; there is no RAM cache, because AVR chips might not have enough RAM for that). Apparently SmartEEPROM already provides a RAM cache for the emulated EEPROM, therefore the read access time should not be a problem.

Reply:
@sigprof So, what needs to be changed to support VIA?

I took a look at https://github.com/qmk/qmk_firmware/pull/6068/files, and it seems it only reads and writes whatever is passed to the function, from the start of the SmartEEPROM, using up to to 36 bytes.

I don't see any memory padding or reallocation happening, and I imagine that would be necessary, so the data is written on a different memory address than VIA data, is that right?

My apologies if all of that sounds basic, I'm pretty green with microcontrollers programming (yet 😁 ), and if I can rewrite the PR to support both preset saving *and* VIA, that would be awesome to do :)

Reply:
VIA support would be indeed pretty awesome and I think it would be a gamechanger 🤩

Reply:
In fact VIA support can already be enabled with current smart eeprom configuration by setting VIA_ENABLE = yes in the rules.mk, and use [alt.json](https://gist.github.com/daltona/23047c4bc94779b9c4f8685acba7cc15)

Reply:
> 
> 
> In fact VIA support can already be enabled with current smart eeprom configuration by setting VIA_ENABLE = yes in the rules.mk, and use [alt.json](https://gist.github.com/daltona/23047c4bc94779b9c4f8685acba7cc15)

Where would the alt.json file go? I tried putting it into my keyboard config folder but no luck (same location as the keymaps folder).

Reply:
Ok I think the alt.json file was for VIA but when I enable using VIA_ENABLE=yes my keyboard does not function and VIA never detects anything

Reply:
You also need to set the EEPROM_SIZE to 1024 in tmk_core/common/arm_atsam/eeprom.c for via to function. Also as indicated by VIA support document, you need to define 4 layers in your keymap
in via you need to enable show design tab in settings tab and load the alt.json file in the design tab.

Reply:
no luck unfortunately. I may just stick to QMK/Massdrops configurator for now until this is hopefully released. Thanks for trying.

Reply:
Ah so I was able to get it working... I believe my issue was not having the right pr merged into my qmk base.

Thanks!

Reply:
@dani190 Do you mind writing a step by step guide of what you did? :)

Reply:
> @dani190 Do you mind writing a step by step guide of what you did? :)

and maybe an binary executable for Windows to get EEPROM and VIA running? Currently I'm using the Drop Configurator on their website. I don't have Visual Studio etc.

Reply:
@thezoc @quaddy I _kinda_ got VIA working on the CTRL thanks to all the contributors awesome work using the following steps:

### Setup mdloader

1. Cloned https://github.com/Massdrop/mdloader
2. Applied #49 (https://github.com/Massdrop/mdloader/pull/49.patch) - `git am 49.patch`
3. Applied #50 (https://github.com/Massdrop/mdloader/pull/50.patch) - `git am 50.patch`
4. Built patched mdloader with VS2019
5. Set SmartEEPROM to the current implementation (I was using #16 before) - `mdloader.exe --first --smarteep --forceeep --restart`

### Build QMK

1. Cloned https://github.com/qmk/qmk_firmware
2. Applied https://github.com/qmk/qmk_firmware/pull/6068 (https://github.com/qmk/qmk_firmware/pull/6068.patch) - `git am 6068.patch`
2. Added a keymap for VIA and applied the `EEPROM_SIZE 1024` mentioned here https://github.com/Massdrop/mdloader/pull/49#discussion_r594894653 (board would not start with the default calculated value) - you can see my commit at https://github.com/kaeltis/qmk_firmware/commit/c90b2b8bf01a56b439ffe813654f46a57817e811
3. Built qmk with `qmk setup kaeltis/qmk_firmware` and `qmk compile -kb massdrop/ctrl -km via`
4. Flashed qmk using `mdloader.exe --first --download massdrop_ctrl_via.bin --restart`

### Use VIA

The steps until now enabled SmartEEPROM in mdloader / the microcontroller, allowed qmk to write to it and also enabled VIA support.

I can now start VIA and click on `File -> Import Keymap` and import a keymap, I tried the one for the ALT from @daltona (https://github.com/Massdrop/mdloader/pull/49#discussion_r590187378) and changed `"productId": "0xeed3"` to `"productId": "0xeed2"` to allow it to detect the CTRL.

In VIA the layout is obviously completely wrong, but it allows me to change keys (just changed everything to the same key for testing) and the board accepts the changes.

The only thing that's missing is the correct keymap for the CTRL, for the ALT the linked one should work.
I tried using http://www.keyboard-layout-editor.com to create one for the CTRL, I however have no idea how the rows/columns are laid out and couldn't find where to look that up.

Reply:
Now I remember that I spent a significant amount of time creating the layout file.
You can find the row column in the source code, the keymaps is organized as ROW, COL so it is a matter of creating the correct layout and filling the row, column accordingly

Reply:
@daltona thats what I was first going to do as well, however according to https://github.com/qmk/qmk_firmware/blob/master/keyboards/massdrop/ctrl/config.h#L31-L33 we have 11 rows and 8 cols - I guess this means something is flipped somewhere.

The second problem is, the keymap as it is in the code has a maximum of 6 rows and 17 cols, which doesn't fit the given rows/cols.

I guess https://github.com/qmk/qmk_firmware/blob/master/keyboards/massdrop/ctrl/ctrl.h#L12-L32 is the actual layout of the rows/cols, I'll try using that as a reference and add some trial and error on top.

Reply:
@kaeltis You're right sorry for the confusion, in fact the 6 rows that I believe existed are in fact split.

to get the matrix layout you should replace all Kxx in the layout view ( the parameters of the macro) with the position in the matrix of the corresponding keycode:

```
0,0 0,1 0,2 0,3 0,4 0,5 0,6 0,7 6,0 6,1 6,2 6,3 6,4 6,5 6,6
6,7 1,0 1,1 1,2 1,3 1,4 1,5 1,6 1,7 7,0 7,1 7,2 7,3 7,4 7,5
etc...
```

Reply:
@daltona thanks! With your help I figured it out :)

Here's the Massdrop CTRL `ctrl.json` for VIA:
https://gist.github.com/kaeltis/c81ffbe79d7e8a92218dd8a610db39b7

Here's the layout on KLE if anyone needs it:
http://www.keyboard-layout-editor.com/#/gists/df6c13bb95885418bf5c6c8f280e93aa